clients/drcachesim/tests/burst_syscall_pt.cpp

```diff
@@ -0,0 +1,240 @@
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+// This application links in drmemtrace_static and acquires a trace during
+// a "burst" of execution that includes some system call traces collected
+// using Intel-PT.
+
+// This is set globally in CMake for other tests so easier to undef here.
+#undef DR_REG_ENUM_COMPATIBILITY
+
+#include "analyzer.h"
+#include "tools/basic_counts.h"
+#include "dr_api.h"
+#include "drmemtrace/drmemtrace.h"
+#include "drmemtrace/raw2trace.h"
+#include "mock_reader.h"
+#include "raw2trace_directory.h"
+#include "scheduler.h"
+
+#include <cassert>
+#include <unistd.h>
+#include <fstream>
+#include <iostream>
+#include <linux/futex.h>
+#include <pthread.h>
+#include <string>
+#include <sys/syscall.h>
+
+namespace dynamorio {
+namespace drmemtrace {
+
+#define FATAL_ERROR(msg, ...)                               \
+    do {                                                    \
+        fprintf(stderr, "ERROR: " msg "\n", ##__VA_ARGS__); \
+        fflush(stderr);                                     \
+        exit(1);                                            \
+    } while (0)
+
+/* The futex the child waits at initially. */
+static uint32_t futex_var = 0xf00d;
+/* The futex the child is transferred to. */
+static uint32_t futex_var_other = 0x8bad;
+
+static void *
+child_futex_wait(void *)
+{
+    long res = syscall(SYS_futex, &futex_var, FUTEX_WAIT, /*#val=*/0xf00d,
+                       /*timeout=*/nullptr, /*uaddr2=*/nullptr, /*val3=*/0);
+    assert(res == 0);
+    std::cerr << "Child released from futex\n";
+    return NULL;
+}
+
+static void
+parent_futex_wake()
+{
+    /* The child would be waiting at the other futex by now.
+     * iX: Note that the child thread undergoes detach while it is waiting
```

Comment:
@derekbruening Thoughts on this? If the process were not under DR, then there would not be any restart for futex and the thread would still be waiting on futex_var_other.

I don't know what the correct native behavior is here. Is a restarted futex supposed to wait on the original var or on the one specified by a possible subsequent FUTEX_CMP_REQUEUE call. I couldn't find documentation in a quick search.

Reply:
I verified using a small test app that the native behavior is that the restarted futex call waits on the original futex, not any futex the thread may have been requeued to using FUTEX_CMP_REQUEUE. So it is indeed a DR transparency violation that our detach signal causes this to happen. Will file a new issue with the native test app.

Reply:
In the example in the futex man page, this would not cause any functional error.  Also, this could happen with any signal, e.g. in an app that uses itimers for self-profiling or something.  So in practice it likely is not significant that we interrupt once for detach, though if it were easy to avoid we would as it is a disturbance.

Reply:
For an app that does not expect any signals, this can be problematic.

> In the example in the futex man page, this would not cause any functional error. 

Am I looking at the correct page (https://man7.org/linux/man-pages/man2/futex.2.html)? I didn't find a code example that uses _REQUEUE.

In any case, the example code at #7034 shows that it indeed causes functional differences. If the app doesn't not unblock the original futex at all (because it assumes the waiting thread should be requeue-d to the other futex), then the requeue-d thread may remain stuck.

Reply:
The man page has this example:
```
                  lock(A)
                  while (!check_value(V)) {
                      unlock(A);
                      block_on(B);
                      lock(A);
                  };
                  unlock(A);
```

The example, and probably most or maybe nearly all, uses of this _REQUEUE flag would be when the target was going to wait on the other futex anyway, and just cleared the first futex.  That's my point: this seems to more likely just affect timing and not correctness.  Yes, you can construct a program that behaves differently for sure, but I bet it's really hard to find that in the wild.